beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogFixedTerm.java

```diff
@@ -0,0 +1,155 @@
+
+        busHandler.pushExpectedEvents(NextEvent.CREATE, NextEvent.BLOCK, NextEvent.NULL_INVOICE);
+        final PlanPhaseSpecifier spec = new PlanPhaseSpecifier("pistol-biennial-inarrear", null);
+        final UUID entitlementId = entitlementApi.createBaseEntitlement(account.getId(), new DefaultEntitlementSpecifier(spec, null, UUID.randomUUID().toString(), null), "something", null, null, false, true, ImmutableList.<PluginProperty>of(), testCallContext);
+        assertNotNull(entitlementId);
+        assertListenerStatus();
+        
+        busHandler.pushExpectedEvents(NextEvent.INVOICE, NextEvent.INVOICE_PAYMENT, NextEvent.PAYMENT); 
+        clock.addYears(2);
+        assertListenerStatus();
+        invoiceChecker.checkInvoice(account.getId(), 1, testCallContext, new ExpectedInvoiceItemCheck(new LocalDate(2021, 12, 01), new LocalDate(2023, 12, 01), InvoiceItemType.RECURRING, new BigDecimal("40.00")));
+        
+        busHandler.pushExpectedEvents(NextEvent.INVOICE, NextEvent.INVOICE_PAYMENT, NextEvent.PAYMENT); 
+        clock.addYears(2);
+        assertListenerStatus();
+        invoiceChecker.checkInvoice(account.getId(), 2, testCallContext, new ExpectedInvoiceItemCheck(new LocalDate(2023, 12, 01), new LocalDate(2024, 12, 01), InvoiceItemType.RECURRING, new BigDecimal("20.03")));
+
+    }
+    
+    @Test(groups = "slow", description = "https://github.com/killbill/killbill/issues/1533")
+    public void testDiscountAndRecurringPhaseInAdvanceBilling() throws Exception {
+
+        busHandler.pushExpectedEvents(NextEvent.CREATE, NextEvent.BLOCK, NextEvent.INVOICE, NextEvent.INVOICE_PAYMENT,
+                                      NextEvent.PAYMENT);
+
+        final PlanPhaseSpecifier spec = new PlanPhaseSpecifier("pistol-biennial-inadvance-discountandrecurring", null);
+        final UUID entitlementId = entitlementApi.createBaseEntitlement(account.getId(), new DefaultEntitlementSpecifier(spec, null, UUID.randomUUID().toString(), null), "something", null, null, false, true, ImmutableList.<PluginProperty>of(), testCallContext);
+        assertNotNull(entitlementId);
+        assertListenerStatus();
+        invoiceChecker.checkInvoice(account.getId(), 1, testCallContext, new ExpectedInvoiceItemCheck(new LocalDate(2021, 12, 01), new LocalDate(2023, 12, 01), InvoiceItemType.RECURRING, new BigDecimal("40.00")));
+
+        busHandler.pushExpectedEvents(NextEvent.INVOICE, NextEvent.INVOICE_PAYMENT, NextEvent.PAYMENT); 
+        clock.addYears(2);
+        assertListenerStatus();
+        invoiceChecker.checkInvoice(account.getId(), 2, testCallContext, new ExpectedInvoiceItemCheck(new LocalDate(2023, 12, 01), new LocalDate(2024, 12, 01), InvoiceItemType.RECURRING, new BigDecimal("20.03")));
+
+        busHandler.pushExpectedEvents(NextEvent.INVOICE, NextEvent.INVOICE_PAYMENT, NextEvent.PAYMENT,NextEvent.PHASE,NextEvent.NULL_INVOICE); //TODO: not sure why NULL_INVOICE is required here but found it in the bus_events_history table
```

Comment:
> //TODO: not sure why NULL_INVOICE is required here but found it in the bus_events_history table

👍 for checking and adding it.

This is because there are 2 'events' triggering an invoice run:
* There is a 'notification' (`notifications` table) with an `effective_date=2024-12-01 12:56:02` that gets triggered as you move the clock and that triggers an invoice run
* There is a bus event (`bus_events` table) for the `PHASE` that also gets triggered as you move the clock and that also triggers an invoice run - to be precise here there is also a notification that first gets processed and as a result triggers this bus event.

If you look at the `notifications` table prior moving the clock you would see: ```select * from notifications;
+-----------+---------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------------+----------------+------------------+---------------------------+------------------+-------------+-------------+-------------+------------------------------------------+---------------------+--------------------------------------+
| record_id | class_name                                                                | event_json                                                                                                                                                                | user_token                           | created_date        | creating_owner | processing_owner | processing_available_date | processing_state | error_count | search_key1 | search_key2 | queue_name                               | effective_date      | future_user_token                    |
+-----------+---------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------------+----------------+------------------+---------------------------+------------------+-------------+-------------+-------------+------------------------------------------+---------------------+--------------------------------------+
|         1 | org.killbill.billing.subscription.engine.core.SubscriptionNotificationKey | {"eventId":"f2e2be15-985a-48e0-9986-12683df2d2ee","seqId":0}                                                                                                              | 409632a6-0485-497c-9acf-b785536abdf1 | 2021-12-01 12:56:03 | Cocodrilo      | NULL             | NULL                      | AVAILABLE        |           0 |           1 |           1 | subscription-service:subscription-events | 2024-12-01 12:56:02 | 64ac62c2-4a25-4de9-85a7-9af8c086f68a |
|         3 | org.killbill.billing.invoice.notification.NextBillingDateNotificationKey  | {"uuidKey":null,"uuidKeys":["4d886e08-6db5-4cf3-a946-5c864370824e"],"targetDate":"2024-12-01T12:56:02.000Z","isDryRunForInvoiceNotification":false,"isRescheduled":false} | bf295f60-0b45-4a6c-b8a9-8bb6573ca2e1 | 2023-12-01 12:56:04 | Cocodrilo      | NULL             | NULL                      | AVAILABLE        |           0 |           1 |           1 | invoice-service:next-billing-date-queue  | 2024-12-01 12:56:02 | 7f366c37-5fa5-4600-8931-97cdfea6139a |
+-----------+---------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------------+----------------+------------------+---------------------------+------------------+-------------+-------------+-------------+------------------------------------------+---------------------+--------------------------------------+
```

So, one of these events will trigger the invoice and the next one will lead to a null invoice.


Note: you can remove your `TODO` if my explanation made sense

Reply:
Understood, thanks for the detailed explanation. I've removed the `TODO`